### ISSUE TITLE
tools/flow: Controller runLoop will halt other running Tasks if one of Tasks failed

### ISSUE THREAD

**Issue by yangzhares** [cue-lang/cue#2313]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
v0.4.3
</pre>

### Does this issue reproduce with the latest stable release?
Yes

### What did you do?

-- tasks.cue -- 
```cue
package main

#Task: {
	id:   "task"
	slow: true | false
}

tasks: {
	slow_task: #Task & {
		slow: true
	}
	fast_task: #Task & {
		slow: false
	}
}
```
-- main.go --
```go
package main

import (
	"context"
	"errors"
	"fmt"
	"os"
	"time"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/tools/flow"
)

// time of executing task
const (
	fastTask = 3
	slowTask = 10
)

func main() {
	ctx := cuecontext.New()
	content, err := os.ReadFile("tasks.cue")
	if err != nil {
		fmt.Printf("read tasks error: %v", err)
		return
	}

	v := ctx.CompileBytes(content, cue.Filename("tasks.cue"))
	if err := v.Err(); err != nil {
		fmt.Printf("compile tasks error: %v", err)
		return
	}

	taskFunc := func(value cue.Value) (flow.Runner, error) {
		idVal := value.LookupPath(cue.ParsePath("id"))
		if !idVal.Exists() {
			return nil, nil
		}

		id, _ := idVal.String()
		if id != "task" {
			return nil, nil
		}

		return flow.RunnerFunc(func(t *flow.Task) error {
			task := t.Value()

			slowVal := task.LookupPath(cue.ParsePath("slow"))
			slow, _ := slowVal.Bool()

			if slow {
				fmt.Println("starting to execute slow task")
				time.Sleep(slowTask * time.Second)
				fmt.Println("completed to execute slow task")
				return nil
			} else {
				fmt.Println("starting to execute fast task")
				time.Sleep(fastTask * time.Second)
				return errors.New("completed to execute fast task")
			}
		}), nil
	}

	ctl := flow.New(&flow.Config{
		Root: cue.ParsePath("tasks"),
	}, v, taskFunc)

	if err := ctl.Run(context.Background()); err != nil {
		fmt.Printf("run tasks error: %v", err)
		return
	}
}

```

-- main.go output --
```
starting to execute fast task
starting to execute slow task
run tasks error: task failed: completed to execute fast task
```

### What did you expect to see?
Expect the slow task already started should be completed even the fast task got failure, in fact, it's halted by `runLoop` from the output. Look at https://github.com/cue-lang/cue/blob/master/tools/flow/run.go#L115-L116, the current logic doesn't care the running tasks completed or not, just simply halted when got the first failure task.

### What did you see instead?
The running task was halted by `runLoop`, don't wait it for completing.

**Comment by myitcv** [comment]
Thanks for raising this, @yangzhares. 

I'm marking this as "working as intended" simply because that's the current semantics of `tools/flow`. 

Yes, we could have opted to not have these fail-fast semantics, but that's what the `tools/flow` package has as of today. 

I assume from you raising this issue that you would like the failure semantics of `tools/flow` to be different/configurable in some way?

**Comment by yangzhares** [comment]
Yes, if it's configurable, that would be better. Maybe a new boolean field like `failFast` added in `flow.Config` that lets have a chance to change current semantics. 

As you said, it's the current semantics, but may i ask if this semantics can be changed via a configurable way? Because the case as I described is a quite common case in everywhere.

**Comment by myitcv** [comment]
> As you said, it's the current semantics, but may i ask if this semantics can be changed via a configurable way?

I don't want to rule it out entirely, but I suspect we will not make changes to the existing API. Instead, we will focus on establishing what the new version of `tools/flow` should look like. That will be a function of https://github.com/cue-lang/cue/issues/1325. Even though that issue is titled `cuerun`, the design of `tools/flow` and `cuerun` are quite linked. Your point is covered in the main description, but please feel free to comment with more specifics.

**Comment by myitcv** [comment]
Closing per discussion above, because this is currently working as intended. And per my comment above this will be addressed/reviewed as part of #1325.
